src/main/scala/Table.scala

```diff
@@ -492,35 +508,71 @@ case class Table private (
     annotations: Map[String, JsonNode],
     var warnings: Array[ErrorWithCsvContext]
 ) {
-  var foreignKeyValues: Map[String, JsonNode] = Map()
 
   /**
     * This array contains the foreign keys defined in other tables' schemas which reference data inside this table.
     */
   var foreignKeyReferences: Array[ForeignKeyWithTable] = Array()
-  var foreignKeyReferenceValues: Map[String, JsonNode] = Map()
-  var warningsFromColumns: Array[ErrorWithCsvContext] =
-    Array[ErrorWithCsvContext]()
-  for (c <- columns) {
-    for (w <- c.warnings) {
-      warningsFromColumns :+= w
-    }
-  }
-  warnings = warnings.concat(warningsFromColumns)
 
-  def validateRow(row: CSVRecord): WarningsAndErrors = {
+  def validateRow(row: CSVRecord): Option[ValidateRowOutput] = {
     var errors = Array[ErrorWithCsvContext]()
+    var primaryKeyValues = List[Any]()
+    var foreignKeyReferenceValues =
+      List[
+        (ForeignKeyWithTable, List[Any])
+      ]() // to store the validated referenced Table Columns values in each row
+    var foreignKeyValues =
+      List[
+        (ForeignKeyWrapper, List[Any])
+      ]() // to store the validated foreign key values in each row
     if (columns.nonEmpty) {
       for ((value, column) <- row.iterator.asScalaArray.zip(columns)) {
         //catch any exception here, possibly outOfBounds  and set warning too many values
-        val result = column.validate(value, row.getRecordNumber)
-        errors = errors.concat(result.errors)
-        warnings = warnings.concat(result.warnings)
+        val (es, v) = column.validate(value)
+        val newValue: List[Any] = if (v.length == 1) {
```

Comment:
Why do we need this if statement? Can't we just do `v.toList` in both cases?

Reply:
not really, I have changed it